torch/csrc/autograd/python_variable.cpp

```diff
@@ -525,16 +525,37 @@ static PyObject* THPVariable_fix_weakref(PyObject* self, PyObject* noargs) {
   Py_RETURN_NONE;
 }
 
+// Maps the given python callable over a vector of items, returning a vector
+// of the same type of items.
+template <typename T>
+static std::vector<T> map_py_func(
+    const py::function& func,
+    const std::vector<T>& items) {
+  std::vector<T> new_items;
+  new_items.resize(items.size());
```

Comment:
This is suboptimal because it will cause default initialization of every element in new_items before you overwrite it. You can use a `back_inserter` for the `new_items.begin()` iterator, in which case you can reserve new_items instead.

GPT-4 wrote this code:

```
template <typename T>
static std::vector<T> map_py_func(
    const py::function& func,
    const std::vector<T>& items) {
  std::vector<T> new_items;
  new_items.reserve(items.size());
  std::transform(
      items.begin(), items.end(), std::back_inserter(new_items), [&](const T& item) {
        return py::cast<T>(func(item));
      });
  return new_items;
}
```

Reply:
I'm scared of this `back_inserter` approach based on what I'm reading [here](https://stackoverflow.com/questions/59284174/is-it-valid-to-use-stdtransform-with-stdback-inserter), but I should be able to avoid the default initialization somehow so I'll keep looking

Reply:
From what I'm reading:
* `std::transform()` + `back_inserter()` doesn't do the right thing for parallel implementations of `transform()`, and it appears C++20 may enable these?
* Using `reserve()` + adding items beyond `.size()` is undefined behavior
* I can use a custom allocator to avoid default initializing when `resize()-ing`, but...

to avoid spending more time here, I'm just going to use a dumb for loop :(